### PR TITLE
Add test for Larmor StateMove throwing if RB is None

### LINKED ISSUE #29414: Exception loading Larmor user files due to beam centre finder scale issue
**Original reporter:** Rob Dalgliesh, @smk78 

### Expected behavior
The ISIS SANS GUI should support all valid user files that v5.0.0 supported

### Actual behavior
There is an unhandled exception loading at least some (possibly all) Larmor user files

### Steps to reproduce the behavior
Must be tested at ISIS:
- Open the ISIS SANS GUI
- Open this user file \\isis\inst$\ndxlarmor\user\users\Masks\User_Larmor_202A_Changer_r51476.txt
- You get an unhandled exception 

### Notes
- I think this is related to the large user file parsing changes from PR #28513 
- The problem is on [this line](https://github.com/mantidproject/mantid/blob/master/scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py#L279) because `data_info.sample_scatter_run_number` is `None`. I'm not sure if it's valid that this is None so this requires investigation to see why that is the case. It seems odd to be checking for a run number before we have specified any runs, so perhaps this is being called by mistake.

### Platforms affected
All

### PR DESCRIPTION
**Pull request by DavidFair**
**Description of work.**

Adds a test to check for StateMove throwing when the run number is none.
This happens when we add a user file without a run number in the box.

I checked with the SANS team and Larmor uses their x dim in
m, and y in mm for runs after 2217. So we want to use the same divider
to keep the display consistent until they load in run data.

**Report to:** @smk78 

**To test:**

- Open the SANS GUI
- Load the following userfile: 
[USER_Larmor_202A_Changer_r51476.txt](https://github.com/mantidproject/mantid/files/5219147/USER_Larmor_202A_Changer_r51476.txt)
- Go to beam centre finder 
- Check centre positions look sane mm values in m form (e.g. 0.01 and 0.001). 
They should not be > 1 (in meters) or tiny (like 1e-5)

Note the display of meters is being done in #29415 

Fixes #29414 

*This does not require release notes* because regression was introduced recently

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py [hunk 1]
```diff
@@ -276,6 +276,13 @@ def __init__(self, data_info):
 
     def _set_conversion_value(self, data_info):
         run_number = data_info.sample_scatter_run_number
+        if not run_number:
+            # We are going through init before file was selected, lets assume > 2217 so
+            # the division is the same as Pos 2 in the case they load a later file
+            # when a file gets loaded this re-runs anyway overwriting our assumptions
+            run_number = 2220
```
Comment (LamarMoore) [inline scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:283]: Are we strictly happy to use a hard-coded value here? Couldn't we just say:
```python
if run_number == None or run_number > 2217:
    self.conversion_value = 1.
```

**Review by LamarMoore — CHANGES_REQUESTED** [review]
Just did a code review. Will do a test on my machine now.

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py (hunk 1 above)
Comment (LamarMoore) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:283]: @DavidFair This is probably an indication that we need to think a bit more about how units are handles so we can shift the cognitive burden off the developer or this will become difficult to develop into.

Comment (DavidFair) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:283]: Agree on folding the if block into a single condition. I was expecting us to have to do something fancy, but it turns out to just be they literally enter one unit as meters and one as mm so I didn't press on how we ended up here

I was saying to Gemma yesterday some sort of set of decorators available to all GUIs which converts from user friendly values into SI units internally could work such as?

```
GuiModel(object):
   def __init__(self):
        self._my_val = 0.012  # In meters

   @property
   @convert_to(units.mm)
   def my_val(self)
       return self._my_val

   @my_val.setter
   @convert_to(units.mm)
   def my_val(self, value)
       return self._my_val
```

Then you would have something like
```
m = GuiModel()
print(m.my_val)  # Returns 0.012 as its internal
print(m.my_val_view)  # <name>_view is generated and correctly applies a scale value to display 12mm

m.my_value_view = view.get_my_value()  # Internally scales the users mm meters then calls the property setter
```


Python doesn't give us much in terms of typing tools without throwing a layer of C++ on top so I imagine this will remove the cognitive burden enough, as exposure to view code tends to be limited

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py [hunk 2]
```diff
@@ -276,7 +276,13 @@ def __init__(self, data_info):
...
     def _set_conversion_value(self, data_info):
         run_number = data_info.sample_scatter_run_number
-        self.conversion_value = 1000. if run_number <= 2217 else 1.
+
+        # If we are going through init before file was selected, lets assume were using
+        # a newer file so in the probable-case the division is the same as Pos 2
+        # When a run number is entered this re-runs anyway overwriting our assumptions
+        # User files after 2217 use Si. units for x but not y
+        if run_number is None or run_number > 2217:
+            self.conversion_value = 1.
```
Comment (LamarMoore) [inline scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:285]: @DavidFair Don't you want to preserve the default of `1000.` if we are on an older file or am I missing something?

Comment (DavidFair) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:285]: We should be fine, this is only called from `__init__` (as it's a builder for the actual object) which already pre-sets it to 1000.

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py (hunk 1 above)
Comment (LamarMoore) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py]: I think unit handling goes beyond just this interface and is perhaps something which needs more careful thought.

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py (hunk 2 above)
Comment (DavidFair) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:285]: Long-term I would like to get rid of the builders and just have the beam centre finder itself (as this is the only place this is used) handle it with a simple check during the running

**Comment by LamarMoore** [comment]
This seems to resolve the issue. We are no longer seeing the unhandled exception when loading the LARMOR user file. Good to go when it is building cleanly.

**Review by LamarMoore — APPROVED** [review]

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py (hunk 2 above)
Comment (LamarMoore) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py:285]: Maybe create an issue

File: scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py (hunk 1 above)
Comment (martyngigg) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py]: I do like the idea of further providing more GUI building blocks though, much like `FileFinderWidget` and if we do a lot of entry where numbers are entered in one unit and convert to another then a widget for handling that is useful for others to build with.

There is additional complication here as the values are dependent on what run number is entered? How do we deal with that in a sane/general manner. Should we? I think there are times when the right answer might be to say we can't deal with everything that might be wrong with input data.
We do have a mechanism for dealing with changing length of [run numbers](https://github.com/mantidproject/mantid/blob/master/instrument/Facilities.xml#L31) but we probably want to be careful at what level we go to allow complete freedom to "fix" bad data.

Comment (DavidFair) [reply scripts/SANS/sans/state/StateObjects/StateMoveDetectors.py]: In this case I would advocate for simplicity rather than edge cases
This problem will go away for SANS too with the migration to TOML files, which partly was to resolve historical weirdness like this

Having some widget which correctly labelled, converted and displayed to a custom precision (to avoid FP rounding errors) would likely get significant usage.
